python/cudf_polars/cudf_polars/experimental/select.py

```diff
@@ -27,10 +57,87 @@ def _(
     if pi.count > 1 and not all(
         expr.is_pointwise for expr in traversal([e.value for e in ir.exprs])
     ):
-        # TODO: Handle non-pointwise expressions.
-        raise NotImplementedError(
-            f"Selection {ir} does not support multiple partitions."
-        )
+        # Try decomposing the underlying expressions
+        return decompose_select(ir, child, partition_info)
+
     new_node = ir.reconstruct([child])
     partition_info[new_node] = pi
     return new_node, partition_info
+
+
+def construct_dataframe(columns: Sequence[Column], names: Sequence[str]) -> DataFrame:
+    """Construct a DataFrame from a sequence of Columns."""
+    return DataFrame(
+        [column.rename(name) for column, name in zip(columns, names, strict=False)]
+    )
+
+
+def build_fusedexpr_select_graph(
+    ir: Select, partition_info: MutableMapping[IR, PartitionInfo]
+) -> MutableMapping[Any, Any]:
+    """Build complex Select graph."""
+    (child,) = ir.children
+    child_name = get_key_name(child)
+    child_count = partition_info[child].count
+
+    # Build Graph to produce a column for each
+    # NamedExpr element of ir.exprs
+    graph: MutableMapping[Any, Any] = {}
+    expr_partition_counts: MutableMapping[Expr, int] = {}
+    roots = []
+    for ne in ir.exprs:
+        assert isinstance(ne.value, FusedExpr), f"{ne.value} is not a FusedExpr"
+        expr: FusedExpr = ne.value
+        roots.append(expr)
+        expr_partition_counts = extract_partition_counts(
+            [expr],
+            child_count,
+            update=expr_partition_counts,
+        )
+        for node in list(traversal([expr]))[::-1]:
+            assert isinstance(node, FusedExpr), f"{node} is not a FusedExpr"
+            graph.update(make_fusedexpr_graph(node, child_name, expr_partition_counts))
+
+    # Add task(s) to select the final columns
+    name = get_key_name(ir)
+    count = max(expr_partition_counts[root] for root in roots)
+    expr_names = [get_key_name(root) for root in roots]
+    expr_bcast = [expr_partition_counts[root] == 1 for root in roots]
+    for i in range(count):
+        graph[(name, i)] = (
+            construct_dataframe,
+            [
+                (name, 0) if bcast else (name, i)
+                for name, bcast in zip(expr_names, expr_bcast, strict=False)
+            ],
+            [ne.name for ne in ir.exprs],
```

Comment:
I think the evaluation of the `FusedExpr` thing should have been wrapped in a `NamedExpr` so that the column name is automatically correct.

Reply:
Yeah, the decomposition makes it so that we cannot wrap each FusedExpr node in a NamedExpr. However, we can keep track of the overarching NamedExpr for each column as we build the graph (and can therefore avoid logic like this). Recent changes should address this.